### ISSUE TITLE
NAN loss or loss gradient when using Binary Cross Entropy or Categorical Cross Entropy sometimes

### ISSUE THREAD

**Issue by manassharma07** [manassharma07/crysx_nn#7]
This is a strange bug, where using a batch_size like 32 or smaller results in nan values in the gradient of loss calculations. But the bug is not there when using a larger batch size like 60-200.

The bug was observed when training on MNIST dataset.

Using ReLU (Hidden, size=256) and Softmax (Output, size=10) activation layers.

**Comment by manassharma07** [comment]
The bug with BCE or Binary Cross Entropy loss seems to have been fixed with the new changes to the code. #

**Comment by manassharma07** [comment]
Actually, the bug may never had been about BCE or CCE, because I just fixed a bug in the Softmax implementation. The maximum for avoiding the over/underflow was supposed to be calculated for each row and not just the whole 2d array.

After fixing the Softmax bug the CCE also works as expected.
